plugins/chain/plugin.cpp

```diff
@@ -352,32 +353,34 @@ namespace chain {
         try {
             ilog("Opening shared memory from ${path}", ("path", my->shared_memory_dir.generic_string()));
             my->db.open(data_dir, my->shared_memory_dir, STEEMIT_INIT_SUPPLY, my->shared_memory_size, chainbase::database::read_write/*, my->validate_invariants*/ );
-            auto head_block_log = my->db.get_block_log().head();
-            if (my->replay_if_struct_changed) {
-                my->replay |= head_block_log && my->db.revision() != head_block_log->block_num();
-            }
-            
-            if (my->replay) {
-                my->replay_db(data_dir, my->force_replay);
-            }
         } catch (const golos::chain::database_revision_exception &) {
-            wlog("Error opening database, attempting to replay blockchain.");
-            my->force_replay |= my->db.revision() >= my->db.head_block_num();
-
-            try {
-                my->replay_db(data_dir, my->force_replay);
-            } catch (const golos::chain::block_log_exception &) {
-                wlog("Error opening block log. Having to resync from network...");
-                my->wipe_db(data_dir, true);
+            if (my->replay_if_corrupted) {
+                wlog("Error opening database, attempting to replay blockchain.");
+                my->force_replay |= my->db.revision() >= my->db.head_block_num();
+                try {
+                    my->replay_db(data_dir, my->force_replay);
+                } catch (const golos::chain::block_log_exception &) {
+                    wlog("Error opening block log. Having to resync from network...");
+                    my->wipe_db(data_dir, true);
+                }
+            } else {
+                wlog("Error opening database, quiting. If should replay, set replay-if-corrupted in config.ini to true.");
+                std::exit(0);
```

Comment:
@afalaleev, why `std::exit(0)`, and not `appbase::app().quit()`?

Reply:
@leinlawun, Tried but causes segfault, it's simpler to just add std::exit(0).

Reply:
Find places in which it is segfaulted and  fix them.

Reply:
@maslenitsa93, and what causes segfault? Tried to look in a debugger?

Reply:
@afalaleev @leinlawun Not yet debugged, but `quit()` don't causes exit in principle, it stops `io_serv` and no more, `shutdown()` not exits too, it just clears plugins. I have no idea what should be use for exit if not raw std, because there is no `appbase.app().abort()` ot some like.

Reply:
@afalaleev @leinlawun `appbase::app().quit()` is just not intented for such cases, it used only inside appbase (and in signal guard). It causes many problems unreasonable to a `medium` issue...
For example, if I trying `shutdown()`+`quit()`+`return`, I have problem in application.hpp:155. If I commenting it out, I have a following segfault in p2p_plugin:
https://pastebin.com/p03GyzmU
If just `quit()`+`return` - following segfault (which one is mentioned above):
https://pastebin.com/LeL48ceG
Why don't just add `abort()` method to appbase which will call `std::exit(0)`? It causes what some code piece not being reached which should? Which?

Reply:
@afalaleev  @leinlawun I debugged all in GDB and I researched how to solve problem:
1) shutdown()+quit()+return
2) In application.cpp, inside application::exec(), should not be called the last `shutdown()`.
My suggestion: create a bool flag in `application` which will store info is already shutted down or no, and if yes - don't do it again. What do you think?

Reply:
We need to shutdown all plugins.
`quit()` looks good.

Reply:
@maslenitsa93, I will make a proactive assumption about the problem: I think the problem is in this line:
https://github.com/GolosChain/golos/blob/2b58fca420b80ddff40179415fdc417147dd0373/plugins/chain/plugin.cpp#L393
`appbase::app().quit()` unloads all plugins and closes the daemon, which causes `plugin::plugin_shutdown()` to be called for each plugin, including the chain plugin, which causes the database to be closed, which is still uninitialized. Is this so or is my guess wrong?

Reply:
@afalaleev Yes. Then:
1. quit()+return
2. Fix problem in `p2p_plugin::plugin_shutdown()`: `my->node` (which is an `unique_ptr`) is null in this case. 
I should add check to plugin_shutdown(): `if (my->node) ...`
It fixes segfault.
Ok?

Reply:
@afalaleev,
https://github.com/GolosChain/golos/blob/04b3e4ada4ba84d12a448b02ba84a0c94edac8f3/plugins/p2p/p2p_plugin.cpp#L502-L503
Is it generally thread-safe?

Reply:
@leinlawun Problem is another, my previous comment is mistake.

Reply:
@afalaleev  @leinlawun Actual problems are in 2 plugins:
1. p2p_plugin plugin_startup calls db::head_block_id() <i>and it does it in another thread</i> (return before it "fixes" it)
2. witness plugin_startup calls db::head_block_num() <i>in <b>same</b> thread</i> (return before it "fixes" it)

Reply:
@afalaleev @leinlawun What if just do not start next plugins if current called quit()?

Reply:
@afalaleev @leinlawun I edited comment.

Reply:
Another ideas:
- don't start plugins but add them to running plugins / or any another way to force shutdown (dirty)
- use initialize-start-initialize-start... instead of first initialize all plugins and next start all plugins

Reply:
@maslenitsa93, In my personal opinion, of course, you need to fixed a witness plugin, and to implement the early completion of the demon, if the previous plugin called the `appbase::app().quit()`. In both cases, an obvious bug.

Reply:
witness_plugin starts mining threads correctly:
https://github.com/GolosChain/golos/blob/771-database-replaying-improvements/plugins/witness/witness.cpp#L188-L194

There are no problems with them.

Reply:
@afalaleev Couldn't we replace their stating to `plugin_startup`?

Reply:
Resizing  of vector with mining threads and start of them are in the one place.
So, if mining_threads exist, they are started. 
This shouldn't take effect on thread.join()

Ok. Let's keep `std::exit(0)` with `TODO comment` for the future, because it is very difficult to check all places which can break.

Reply:
@afalaleev Okay.